# Air Quality Index for Cities in India 2015 - 2020

In [2]:
#import libraries
import pandas as pd 
import numpy as np

In [3]:
#load data 
india_air_data = pd.read_csv("/home/aavu/Documents/Project-Data/AQI/Data/city_day.csv")
india_air_data.head()

City        Date  PM2.5  PM10     NO    NO2    NOx  NH3     CO    SO2  \
0  Ahmedabad  2015-01-01    NaN   NaN   0.92  18.22  17.15  NaN   0.92  27.64   
1  Ahmedabad  2015-01-02    NaN   NaN   0.97  15.69  16.46  NaN   0.97  24.55   
2  Ahmedabad  2015-01-03    NaN   NaN  17.40  19.30  29.70  NaN  17.40  29.07   
3  Ahmedabad  2015-01-04    NaN   NaN   1.70  18.48  17.97  NaN   1.70  18.59   
4  Ahmedabad  2015-01-05    NaN   NaN  22.10  21.42  37.76  NaN  22.10  39.33   

       O3  Benzene  Toluene  Xylene  AQI AQI_Bucket  
0  133.36     0.00     0.02    0.00  NaN        NaN  
1   34.06     3.68     5.50    3.77  NaN        NaN  
2   30.70     6.80    16.40    2.25  NaN        NaN  
3   36.08     4.43    10.14    1.00  NaN        NaN  
4   39.31     7.01    18.89    2.78  NaN        NaN

In [4]:
india_air_data.tail()

City        Date  PM2.5   PM10    NO    NO2    NOx    NH3  \
29526  Visakhapatnam  2020-06-27  15.02  50.94  7.68  25.06  19.54  12.47   
29527  Visakhapatnam  2020-06-28  24.38  74.09  3.42  26.06  16.53  11.99   
29528  Visakhapatnam  2020-06-29  22.91  65.73  3.45  29.53  18.33  10.71   
29529  Visakhapatnam  2020-06-30  16.64  49.97  4.05  29.26  18.80  10.03   
29530  Visakhapatnam  2020-07-01  15.00  66.00  0.40  26.85  14.05   5.20   

         CO    SO2     O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
29526  0.47   8.55  23.30     2.24    12.07    0.73  41.0          Good  
29527  0.52  12.72  30.14     0.74     2.21    0.38  70.0  Satisfactory  
29528  0.48   8.42  30.96     0.01     0.01    0.00  68.0  Satisfactory  
29529  0.52   9.84  28.30     0.00     0.00    0.00  54.0  Satisfactory  
29530  0.59   2.10  17.05      NaN      NaN     NaN  50.0          Good

In [5]:
india_air_data['AQI_Bucket'].unique()

array([nan, 'Poor', 'Very Poor', 'Severe', 'Moderate', 'Satisfactory',
       'Good'], dtype=object)

In [6]:
#data description 
india_air_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [7]:
india_air_data.shape

(29531, 16)

In [8]:
print(india_air_data['City'].nunique())
print(india_air_data['City'].unique())

26
['Ahmedabad' 'Aizawl' 'Amaravati' 'Amritsar' 'Bengaluru' 'Bhopal'
 'Brajrajnagar' 'Chandigarh' 'Chennai' 'Coimbatore' 'Delhi' 'Ernakulam'
 'Gurugram' 'Guwahati' 'Hyderabad' 'Jaipur' 'Jorapokhar' 'Kochi' 'Kolkata'
 'Lucknow' 'Mumbai' 'Patna' 'Shillong' 'Talcher' 'Thiruvananthapuram'
 'Visakhapatnam']


In [9]:
pollutant_columns = india_air_data[['PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2','O3', 'Benzene', 'Toluene', 'Xylene', 'AQI']]
pollutant_columns.describe()

PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775     21.694928     15.811136   
min        0.010000      0.000000      0.010000      0.010000      0.000000   
25%        8.580000      0.510000      5.670000     18.860000      0.120000   
50%       15.850000      0.890000      9.160000     30.840000      1.070000   
75%       30.020000      1.450000     15.220000     45.570000      3.080000   
max      352.890000    175.810000    193.860000    257.730000    455.030000   

            Toluene        Xylene           AQI  
count  21490.000000  11422.000000  24850.000000  
mean       8.700972      3.070128    166.463581  
std       19.969164      6.323247    140.696585  
min        0.000000      0.000000     13.000000  
25%        0.600000      0.140000     81.000000  
50%        2.970000      0.980000    118.000000  
75%        9.150000      3.350000    208.000000  
max      454.850000    170.370000   2049.000000

In [10]:
#clean data 
missing_values_count = india_air_data.isnull().sum()
total_cells = np.product(india_air_data.shape)
total_missing = missing_values_count.sum()
print(missing_values_count[:])

City              0
Date              0
PM2.5          4598
PM10          11140
NO             3582
NO2            3585
NOx            4185
NH3           10328
CO             2059
SO2            3854
O3             4022
Benzene        5623
Toluene        8041
Xylene        18109
AQI            4681
AQI_Bucket     4681
dtype: int64


In [11]:
import statistics

In [12]:
#Replace the missing values

india_air_data['PM2.5'].fillna(india_air_data['PM2.5'].mean(),inplace = True )
india_air_data['PM10'].fillna(india_air_data['PM10'].mean(),inplace = True )
india_air_data['NO'].fillna(india_air_data['NO'].mean(),inplace = True )
india_air_data['NO2'].fillna(india_air_data['NO2'].mean(),inplace = True )
india_air_data['NOx'].fillna(india_air_data['NOx'].mean(),inplace = True )
india_air_data['NH3'].fillna(india_air_data['NH3'].mean(),inplace = True )
india_air_data['CO'].fillna(india_air_data['CO'].mean(),inplace = True )
india_air_data['SO2'].fillna(india_air_data['SO2'].mean(),inplace = True )
india_air_data['O3'].fillna(india_air_data['O3'].mean(),inplace = True )
india_air_data['Benzene'].fillna(india_air_data['Benzene'].mean(),inplace = True )
india_air_data['Toluene'].fillna(india_air_data['Toluene'].mean(),inplace = True )
india_air_data['Xylene'].fillna(india_air_data['Xylene'].mean(),inplace = True )
india_air_data['AQI'].fillna(india_air_data['AQI'].mean(),inplace = True )

In [13]:
india_air_data['AQI_Bucket'].mode()

0    Moderate
dtype: object

In [14]:
india_air_data['AQI_Bucket'].isnull().sum()

4681

In total, there are 4,681 missing descriptions in the 'AQI Bucket' column. Let's populate the missing rows in accordance with the score system implemented. 

AQI Score System: -
* Good (0-50)
* Satisfactory (51-100)
* Moderate (101-200)
* Poor (201-300)
* Very poor (301-400)
* Severe (401-500)

In [15]:
india_air_data.loc[(india_air_data['AQI'] >= 0) & (india_air_data['AQI'] <= 50), 'AQI_Bucket'] = 'Good'
india_air_data.loc[(india_air_data['AQI'] >= 51) & (india_air_data['AQI'] <= 100), 'AQI_Bucket'] = 'Satisfactory'
india_air_data.loc[(india_air_data['AQI'] >= 101) & (india_air_data['AQI'] <= 200), 'AQI_Bucket'] = 'Moderate'
india_air_data.loc[(india_air_data['AQI'] >= 201) & (india_air_data['AQI'] <= 300), 'AQI_Bucket'] = 'Poor'
india_air_data.loc[(india_air_data['AQI'] >= 301) & (india_air_data['AQI'] <= 400), 'AQI_Bucket'] = 'Very poor'
india_air_data.loc[(india_air_data['AQI'] >= 401) & (india_air_data['AQI'] <= 500), 'AQI_Bucket'] = 'Severe'

In [16]:
#check if null values filled correctly 
india_air_data['AQI_Bucket'].isnull().sum()

0

In [17]:
india_air_data.corr()

PM2.5      PM10        NO       NO2       NOx       NH3        CO  \
PM2.5    1.000000  0.558079  0.426375  0.344341  0.380725  0.189227  0.086663   
PM10     0.558079  1.000000  0.431006  0.359165  0.415133  0.223025  0.047517   
NO       0.426375  0.431006  1.000000  0.462402  0.746223  0.156394  0.211639   
NO2      0.344341  0.359165  0.462402  1.000000  0.574190  0.165984  0.353237   
NOx      0.380725  0.415133  0.746223  0.574190  1.000000  0.128051  0.225097   
NH3      0.189227  0.223025  0.156394  0.165984  0.128051  1.000000  0.020029   
CO       0.086663  0.047517  0.211639  0.353237  0.225097  0.020029  1.000000   
SO2      0.119512  0.176188  0.166190  0.382758  0.208355 -0.021005  0.472583   
O3       0.155330  0.203595  0.014218  0.285448  0.083063  0.078688  0.039787   
Benzene  0.021934  0.019215  0.033901  0.025082  0.037383 -0.011864  0.061351   
Toluene  0.107788  0.121983  0.134201  0.254074  0.168780  0.007442  0.274882   
Xylene   0.070459  0.031256  0.059494  0.133037  0.056920 -0.002215  0.145190   
AQI      0.628860  0.484497  0.430600  0.522994  0.438363  0.137436  0.649679   

              SO2        O3   Benzene   Toluene    Xylene       AQI  
PM2.5    0.119512  0.155330  0.021934  0.107788  0.070459  0.628860  
PM10     0.176188  0.203595  0.019215  0.121983  0.031256  0.484497  
NO       0.166190  0.014218  0.033901  0.134201  0.059494  0.430600  
NO2      0.382758  0.285448  0.025082  0.254074  0.133037  0.522994  
NOx      0.208355  0.083063  0.037383  0.168780  0.056920  0.438363  
NH3     -0.021005  0.078688 -0.011864  0.007442 -0.002215  0.137436  
CO       0.472583  0.039787  0.061351  0.274882  0.145190  0.649679  
SO2      1.000000  0.156610  0.033059  0.265522  0.203766  0.452768  
O3       0.156610  1.000000  0.018748  0.113683  0.068016  0.188590  
Benzene  0.033059  0.018748  1.000000  0.694691  0.092957  0.041518  
Toluene  0.265522  0.113683  0.694691  1.000000  0.255347  0.262522  
Xylene   0.203766  0.068016  0.092957  0.255347  1.000000  0.139934  
AQI      0.452768  0.188590  0.041518  0.262522  0.139934  1.000000

In [18]:
missing_values_count = india_air_data.isnull().sum()
total_cells = np.product(india_air_data.shape)
total_missing = missing_values_count.sum()
print(missing_values_count[:])

City          0
Date          0
PM2.5         0
PM10          0
NO            0
NO2           0
NOx           0
NH3           0
CO            0
SO2           0
O3            0
Benzene       0
Toluene       0
Xylene        0
AQI           0
AQI_Bucket    0
dtype: int64
